### DISCUSSION TITLE
AckWait does not publish to dead letter queue with zero clients

### DISCUSSION THREAD

**Discussion by gongy** [nats-io/nats-server#4994] [Q&A] [answered]
We're looking into using NATS JetStream to manage a large number of work queues, each with a durable consumer, polled from a client pool which scales to zero.

The behavior explored in this [issue](https://github.com/nats-io/nats-server/issues/1716) confirms that intentionally, if `AckWait` expires with no clients fetching from the pull consumer, the message does not move to the dead letter queue (`MaxDeliver=1`). We have also reproduced this effect, and understand it is by design.

There appears to be no way to configure JetStream to eagerly move messages from pending to the dead letter queue upon `AckWait` timeout, without clients issuing fetches. The behavior of a push consumer + queue group setup appears to exhibit the same effect - if there are no subscribers in the queue group, messages remain in `AckPending` after the `AckWait`.

One concept we explored is sending dummy fetch requests to the pull consumer, to activate the message redirection to the dead letter queue. We did not find a way to trigger this manually, however.

Would you have any suggestions for a work-around that allows for scale-to-zero client pools for pull consumers?

**Comment by derekcollison** [comment, marked as answer]
We do not have automated DLQs, as you noted by design since we truly separate consumers from stream semantics.

Also note in your case you can run one large mux'd stream and have consumer pull only the subject they need, most likely bound to the type of event.

For DLQ, you could term a message and have something that captures that advisory and moves the message to whichever stream you want.
